[Week5] Add validation to the cat toy store And Create member management

JamieShin0201 opened the pull request:
# 해야 할 일
- [X] DozerMapper 추가 및 적용
- [X] 회원 생성하기 - POST /user
- [X] 회원 수정하기 - PATCH /user/{id}
- [X] 회원 삭제하기 - DELETE /user/{id}
- [ ] 전체적으로 살펴보고 리팩토링하기

## 요구 사항
- [X] 테스트 커버리지 100%를 달성해야 합니다.
- [X] 모든 API 테스트를 통과해야 합니다.
- [X] 모든 E2E 테스트를 통과해야 합니다.

JamieShin0201 committed:
Refactor ProductControllerTest With Bdd Style

JamieShin0201 committed:
Add DozerMapper

Added DozerMapper's Dependency
Added DozerMapper bean

JamieShin0201 committed:
Apply DozerMapper to ProductService

JamieShin0201 committed:
Add @CrossOrigin in ProductController

JamieShin0201 committed:
Clean unnecessary lines, imports

JamieShin0201 committed:
[Red, Green] Add User, UserTest

JamieShin0201 committed:
[Red] Add getUsers test in UserServiceTest

Added UserService
Added UserRepository
Added JpaUserRepository

JamieShin0201 committed:
[Green] Implement getUsers in UserService

JamieShin0201 committed:
[Green] Add getUsers method's empty user list test in UserServiceTest

JamieShin0201 committed:
[Red] Add getUser test in UserServiceTest

JamieShin0201 committed:
[Green] Implement getUser in UserService

JamieShin0201 committed:
[Red] Add getUser with not existing id test in UserServiceTest

Added UserNotFoundException

JamieShin0201 committed:
[Refactor] Refactor getUser in UserService

JamieShin0201 committed:
[Red] Add updateUser test in UserServiceTest

Added DTO UserUpdateRequest

JamieShin0201 committed:
[Green] Implement updateUser in UserService

Added changeWith method in User
Added dozer mapper in UserService, UserServiceTest
Added mapper in UserServiceTest

JamieShin0201 committed:
[Refactor] Add findUser method in UserService

JamieShin0201 committed:
[Red] Add deleteUser test in UserServiceTest

JamieShin0201 committed:
[Green] Add deleteUser in UserService

JamieShin0201 committed:
[Red] Add createUser test in UserServiceRepository

Added DTO UserCreateRequest
Added save method in repository

JamieShin0201 committed:
[Green] Implement createUser in UserService

JamieShin0201 committed:
[Red] GET /users test

JamieShin0201 committed:
[Green] Implement GET /users

JamieShin0201 committed:
[Red] GET /users/{id} test in UserControllerTest

JamieShin0201 committed:
[Green] Implement GET /users/{id}

Added handleUserNotFound method in NotFoundErrorAdvice

JamieShin0201 committed:
[Red] Add POST /users test in UserControllerTest

JamieShin0201 committed:
[Green] Implement POST /users

JamieShin0201 committed:
[Red] Add Patch /users/{id} test in UserControllerTest

JamieShin0201 committed:
[Green] Implement PATCH /users/{id}

JamieShin0201 committed:
[Red] Add DELETE /users/{id} test

JamieShin0201 committed:
[Green] Imeplement Delete /users/{id}

JamieShin0201 committed:
[Refactor] Change getUsers's return type to UserReponse

JamieShin0201 committed:
Replace deleteuser method in UserService

JamieShin0201 committed:
Organize folders

JamieShin0201 committed:
Add updateUser test in UserTest

JamieShin0201 committed:
Add JavaDoc in User

JamieShin0201 committed:
Add JavaDoc in UserController

JamieShin0201 committed:
Add JavaDoc in UserService

JamieShin0201 commented:
# 2021/02/23 작업내용
1. ProductService에 DozerMapper를 적용하였습니다.
2. 회원 관련된 요구사항들을 구현하였습니다. (생성, 수정, 삭제)
3. 회원 엔터티가 외부에 노출되지 않도록 DTO를 생성했습니다.
4. 회원과 상품 폴더를 생성해 전반적으로 클래스 위치를 정리하였습니다.

TDD로 진행하는중 빠르게 성공하고 리팩토링해야하는데.. 구현할때 리팩토링도 생각하면서 구현을 하게 되네요.ㅠㅠ
다시 연습해봐야 할 것 같습니다.!

추가적으로 코드 살펴보면서 리팩토링 하겠습니다!.!

감사합니다^_^

hannut91 commented on app/src/main/java/com/codesoom/assignment/user/service/UserService.java:
```diff
@@ -0,0 +1,95 @@
+    /**
+     * 모든 회원을 리턴합니다.
+     */
+    public List<UserResponse> getUsers() {
+        List<User> users = userRepository.findAll();
+
+        return users.stream()
+                .map(user -> mapper.map(user, UserResponse.class))
+                .collect(Collectors.toList());
+    }
```
서비스에서 바깥의 Controller의 UserResponse에 의존하고 있는 것 같습니다.

hannut91 commented on app/src/test/java/com/codesoom/assignment/user/service/UserServiceTest.java:
```diff
@@ -0,0 +1,229 @@
+            @Test
+            @DisplayName("수정된 회원을 리턴한다.")
+            void it_returns_the_updated_user() {
+                UserResponse updatedUser = userService.updateUser(existingId, userUpdateRequest);
+
+                assertThat(updatedUser.getName()).isEqualTo(userUpdateRequest.getName());
```
단어들이 굉장히 비슷해서 읽기가 어려운 것 같습니다. update, user 이 순서만 바뀌다보니 헷갈리는 것 같아요

그리고 userService가 UserResponse를 반환하는 것이 이해하기 어려운 것 같습니다. 사용자를 수정해는데, 사용자 응답이 반환되는 것이니 헷갈리는 것 같아요.

```java
User user = userService.updateUser(existingId, userUpdateRequest);

assertThat(user.getName()).isEqualTo(userUpdateRequest.getName());
```

hannut91 commented on app/src/test/java/com/codesoom/assignment/user/service/UserServiceTest.java:
```diff
@@ -0,0 +1,229 @@
+        @Test
+        @DisplayName("생성된 회원을 리턴한다.")
+        void it_return_the_created_user() {
+            UserResponse createdUser = userService.createUser(userCreateRequest);
```
한 문장에 이 createdUser, createUser, userCreateRequset 세 단어들이 나옵니다. 비슷한 모양의 단어들을 사용하는 것은 별로 안좋을 것 같아요

See also
  - https://ko.javascript.info/ninja-code#ref-2033

hannut91 commented:
바쁜 일이 있으셨군요 ㅎㅎ 도움이 필요하시면 말씀해주세요

JamieShin0201 committed:
Refactor UserController, UserService

locate mapper in UserController

'#' will be ignored, and an empty message aborts the commit.

JamieShin0201 commented:
# 2021/02/25 작업내용
1. Dto에 대한 의존성을 Service에서 Controller로 변경하였습니다.
2. 비슷한 모양의 단어들을 수정하였습니다.

오늘 작업량이 적네요 ㅠ.ㅠ 
계속 코드를 살펴보면서 수정하겠습니다!.!

감사합니다^_^

JamieShin0201 committed:
Add throws's JavaDoc in UserService

JamieShin0201 committed:
Make sure that password is not responded

JamieShin0201 committed:
Remove builder in UserResponse

JamieShin0201 committed:
Add @Email for checking vaild email

JamieShin0201 committed:
Make email to unique

hannut91 commented on app/src/test/java/com/codesoom/assignment/user/service/UserServiceTest.java:
```diff
@@ -0,0 +1,227 @@
+            @Test
+            @DisplayName("찾은 회원을 리턴한다.")
+            void it_returns_the_found_user() {
+                User foundUser = userService.getUser(existingId);
```
```suggestion
                User foundUser = userService.getUser(user1.getId());
```

이렇게 하는 것은 어떨까요?

JamieShin0201 committed:
Refactor UserServiceTest

JamieShin0201 committed:
Handle DuplicateUserException

JamieShin0201 committed:
Handle DuplicateUserException in UserController

JamieShin0201 commented:
# 2021/02/26 작업내용
1. email이 중복되는 경우 회원생성이 되지 않도록 수정하였습니다.
2. 리뷰 받은부분을 수정하였습니다.

감사합니다^_^

hannut91 commented on app/src/main/java/com/codesoom/assignment/common/exceptions/UserNotFoundException.java:
```diff
@@ -0,0 +1,13 @@
+
+public class UserNotFoundException extends RuntimeException {
+
+    public UserNotFoundException() {
+        super("회원을 찾을 수 없습니다.");
+    }
+
+    public UserNotFoundException(String message) {
+        super(message);
+    }
```
아래 생성자가 주 생성자라면, 위의 생성자는 부 생성자가 될 것이고 그러면 사실 다음과 같이 되는게 좋겠네요.

```suggestion
    public UserNotFoundException() {
        this("회원을 찾을 수 없습니다.");
    }

    public UserNotFoundException(String message) {
        super(message);
    }
```

hannut91 commented on app/src/main/java/com/codesoom/assignment/user/controller/UserController.java:
```diff
@@ -0,0 +1,107 @@
+     * @param id 찾고자 하는 회원의 식별자
+     * @return 찾은 회원
+     */
+    @GetMapping("{id}")
+    public UserResponse find(@PathVariable Long id) {
+        User user = userService.getUser(id);
+
+        return mapper.map(user, UserResponse.class);
+    }
```
`find`는 여러개를 찾을 것 같은데, `findOne`같은 것은 어떨까요? 너무 database에서 쓸 것만 같은 이름일까요?

JamieShin0201 replied:
findOne으로 해도 괜찮을 것 같습니다. 혹은 findUser도 괜찮을 것 같습니다!

JamieShin0201 committed:
Fix UserNotFoundException

JamieShin0201 committed:
Fix DuplicateUserException

JamieShin0201 committed:
Fix find -> findOne in UserController

hannut91 reviewed (approved):
회사를 다니면서 과제를 하는 것만으로도 대단하다고 생각이듭니다 ㅎㅎ 그리고 코드숨 스터디고 같이 하고 계시잖아요. 얼마나 공부할지 결정하는 힘이 필요할 것 같습니다.

고생하셨습니다!
